ruby-sinatra/src/main.rb

```diff
@@ -0,0 +1,317 @@
+    set :database, DATABASE_URL
+  else
+    # Fall back to specific configurations for development and test environments
+    case settings.environment
+    when :development
+      set :database, { adapter: 'postgresql', database: 'waect' }
+      ActiveRecord.verbose_query_logs = true
+    when :test
+      set :database, { adapter: 'postgresql', database: 'minitwit_test' }
+      enable :logging
+      ActiveRecord::Base.logger = Logger.new($stdout)
+    end
+  end
+  set :public_folder, "#{__dir__}/static"
+  enable :sessions
+end
+
+helpers do
+  def logged_in?
+    !!session[:user_id]
+  end
+
+  def current_user
+    return unless logged_in?
+
+    @current_user ||= User.find(session[:user_id])
+  end
+
+  def request_is_not_from_simulator
+    request.env['HTTP_AUTHORIZATION'] != 'Basic c2ltdWxhdG9yOnN1cGVyX3NhZmUh'
+  end
+end
+
+def update_latest(request)
+  parsed_command_id = request.params['latest']
+  Latest.set(parsed_command_id.to_i) if parsed_command_id.present?
+end
+
+before '/api/*' do
+  update_latest(request)
+  if request_is_not_from_simulator
+    halt [403, { status: 403, error_msg: 'You are not authorized to use this resource!' }.to_json]
+  end
+end
+
+get '/' do
+  redirect('/public') unless logged_in?
+
+  @title = 'My Timeline'
+
+  @messages = Message
+              .unflagged
+              .authored_by(current_user.following + [current_user])
+              .includes(:author)
+              .order(pub_date: :desc)
+              .first(PR_PAGE)
+
+  erb :timeline, layout: :layout
+end
+
+get '/public' do
+  @title = 'Public Timeline'
+  @messages = Message
+              .unflagged
+              .includes(:author)
+              .order(pub_date: :desc)
+              .first(PR_PAGE)
+
+  erb :timeline, layout: :layout
+end
+
+get '/register' do
+  erb :register, layout: :layout
+end
+
+post '/register' do
+  user = User.new(
+    username: params[:username],
+    email: params[:email],
+    password: params[:password],
+    password_confirmation: params[:password2]
+  )
+  if user.save
+    flash[:success] = 'You were successfully registered and can login now'
+    redirect('/login')
+  else
+    errors = user.errors.map(&:full_message).join(', ')
+    flash[:error] = errors
+  end
+  redirect('/register')
+end
+
+get '/login' do
+  erb :login, layout: :layout
+end
+
+post '/login' do
+  user = User.find_by_username(params[:username])
+  if user.nil?
+    error = 'Invalid username'
+  elsif !user.authenticate(params[:password])
+    error = 'Invalid password'
+  else
+    session[:user_id] = user.id
+    flash[:success] = 'You were logged in'
+    redirect('/')
+  end
+  flash[:error] = error
+  redirect('/login')
+end
+
+get '/logout' do
+  session[:user_id] = nil
+  flash[:success] = 'You were logged out'
+  redirect('/')
+end
+
+post '/add_message' do
+  if !session[:user_id]
+    return status 401
+  elsif params[:text].nil? || params[:text].strip.empty?
+    flash[:error] = 'Message cannot be empty!'
```

Comment:
The other applications do not have this flash-message. Go-Gin have something similar but a different message. We should align.

Reply:
yep, we need to allign the flash errors. But honestly, since we are not expecting to force errors it shouldnt be a difference during the scenarios execution.